flink-connectors/flink-connector-rabbitmq2/src/main/java/org/apache/flink/connector/rabbitmq2/source/enumerator/RabbitMQSourceEnumerator.java

```diff
@@ -0,0 +1,120 @@
+
+import org.apache.flink.api.connector.source.SplitEnumerator;
+import org.apache.flink.api.connector.source.SplitEnumeratorContext;
+import org.apache.flink.api.connector.source.SplitsAssignment;
+import org.apache.flink.connector.rabbitmq2.ConsistencyMode;
+import org.apache.flink.connector.rabbitmq2.RabbitMQConnectionConfig;
+import org.apache.flink.connector.rabbitmq2.source.split.RabbitMQSourceSplit;
+import org.apache.flink.util.FlinkRuntimeException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.util.List;
+
+/**
+ * The source enumerator provides the source readers with the split. All source readers receive the
+ * same split as it only contains information about the connection and in case of exactly-once, the
+ * seen correlation ids. But in this case, the enumerator makes sure that at maximum one source
+ * reader receives the split. During exactly-once if multiple reader should be assigned a split a
+ * {@link FlinkRuntimeException} is thrown.
+ */
+public class RabbitMQSourceEnumerator
+        implements SplitEnumerator<RabbitMQSourceSplit, RabbitMQSourceEnumState> {
+    private final SplitEnumeratorContext<RabbitMQSourceSplit> context;
+    private final ConsistencyMode consistencyMode;
+    private static final Logger LOG = LoggerFactory.getLogger(RabbitMQSourceEnumerator.class);
+    private RabbitMQSourceSplit split;
+
+    public RabbitMQSourceEnumerator(
+            SplitEnumeratorContext<RabbitMQSourceSplit> context,
+            ConsistencyMode consistencyMode,
+            RabbitMQConnectionConfig connectionConfig,
+            String rmqQueueName,
+            RabbitMQSourceEnumState enumState) {
+        // The enumState is not used since the enumerator has no state in this architecture.
+        this(context, consistencyMode, connectionConfig, rmqQueueName);
+    }
+
+    public RabbitMQSourceEnumerator(
+            SplitEnumeratorContext<RabbitMQSourceSplit> context,
+            ConsistencyMode consistencyMode,
+            RabbitMQConnectionConfig connectionConfig,
+            String rmqQueueName) {
+        this.context = context;
+        this.consistencyMode = consistencyMode;
+        this.split = new RabbitMQSourceSplit(connectionConfig, rmqQueueName);
+    }
+
+    @Override
+    public void start() {
+        LOG.info("Start rabbitmq source enumerator");
+    }
+
+    @Override
+    public void handleSplitRequest(int i, @Nullable String s) {
+        LOG.info("Split request from reader " + i);
+        assignSplitToReader(i, split);
+    }
+
+    @Override
+    public void addSplitsBack(List<RabbitMQSourceSplit> list, int i) {
+        LOG.info("Splits returned from reader " + i);
+        if (list.size() == 0) {
+            return;
+        }
+        // Every Source Reader will only receive one splits, thus we will never get back more.
+        assert list.size() == 1;
+        split = list.get(0);
```

Comment:
In case of exactly-once the single split gets updated with the correlation ids and therefore needs to be updated in the enumerator. For the others, you are right, it doesn't matter. We kept the exactly-once case to simplify the code. Do you have a suggestion to make it more clear?

Reply:
Done.